fms/triton/moe_kernel.py

```diff
@@ -0,0 +1,419 @@
+
+
+def _load_best_configs():
+    saved_configs = pathlib.Path.cwd() / "moe_mm_configs.p"
+    if saved_configs.is_file():
+        import pickle
+
+        with open(saved_configs, "rb") as f:
+            print(f"Loading best configs from file {saved_configs}")
+            return pickle.load(f)
+
+
+def _save_best_configs(best_configs):
+    saved_configs = pathlib.Path.cwd() / "moe_mm_configs.p"
+    with open(saved_configs, "wb") as f:
+        import pickle
+
+        print(f"Saving best configs to file {saved_configs}")
+        pickle.dump(best_configs, f)
+
+
+def _create_best_configs_key(input, moe_matrix, token_expert_mapping):
+    key = (input.size(), token_expert_mapping.size(), moe_matrix.size())
+    return key
+
+
+BEST_MOE_CONFIGS = None
+
+lib = torch.library.Library("moe", "FRAGMENT")
+lib.define(
+    "moe_mm(Tensor input, Tensor moe_matrix, Tensor token_expert_mapping, Tensor padded_token_ids_per_block, Tensor expert_block_mapping, Tensor total_padded_tokens, int topk, int padding_size) -> Tensor"
+)
+
+
+# All that's needed for torch.compile support
+@torch.library.impl(lib, "moe_mm", "Meta")
+def moe_mm_meta(
+    input: torch.Tensor,
+    moe_matrix: torch.Tensor,
+    token_expert_mapping: torch.Tensor,
+    padded_token_ids_per_block: torch.Tensor,
+    expert_block_mapping: torch.Tensor,
+    total_padded_tokens: torch.Tensor,
+    topk: int,
+    padding_size,
+):
+    M, A = token_expert_mapping.shape
+    _, N, _ = moe_matrix.shape
+    return torch.empty((M, A, N), device=input.device, dtype=input.dtype)
+
+
+@torch.library.impl(lib, "moe_mm", "CUDA")
+def moe_mm(
+    input: torch.Tensor,
+    moe_matrix: torch.Tensor,
+    token_expert_mapping: torch.Tensor,
+    padded_token_ids_per_block: torch.Tensor,
+    expert_block_mapping: torch.Tensor,
+    total_padded_tokens: torch.Tensor,
+    topk: int,
+    padding_size,
+):
+    M, A = token_expert_mapping.shape
+    E, N, _ = moe_matrix.shape
+    output = torch.zeros((M, A, N), device=input.device, dtype=input.dtype)
+
+    global BEST_MOE_CONFIGS
+    if BEST_MOE_CONFIGS is None:
+        BEST_MOE_CONFIGS = _load_best_configs()
+    # Loading must have not been successful. Let's create a new dictionary.
+    if BEST_MOE_CONFIGS is None:
+        BEST_MOE_CONFIGS = {}
+    key = _create_best_configs_key(input, moe_matrix, token_expert_mapping)
+    if key not in BEST_MOE_CONFIGS:
+        print("key ", key, " not found. Running autotune. This might take a while.")
+        import functools
+
+        # TODO: Add more configs?
+        configs = [
+            {
+                "block_m": 64,
+                "block_n": 64,
+                "block_k": 32,
+            },
+            {
+                "block_m": 16,
+                "block_n": 32,
+                "block_k": 64,
+            },
+        ]
+
+        configs = [config for config in configs if config["block_m"] == padding_size]
+        print("all configs len: ", len(configs))
```

Comment:
I'm thinking you probably didn't mean to leave in this (or the below) print statements